bitwarden_access_manager.py

```diff
@@ -0,0 +1,672 @@
+
+                for member_id in member_ids:
+                    data = self._http_get(f"/public/members/{member_id}")
+                    member = json.load(data)
+                    yield MemberCollectionAccess(
+                        name=member["name"],
+                    )
+
+    def get_collection_groups(self, groups: Any) -> Iterable[GroupCollectionAccess]:
+
+            for group in groups:
+                if group["readOnly"] == True:
+                    access = "readonly"
+                else:
+                    access = "write"
+
+                yield GroupCollectionAccess(
+                    name=self.get_group(group["id"])["name"],
+                    access=access,
+                )
+
+    def get_member_collections(self, members: Tuple[MemberCollectionAccess, ...]):
+        print("JERE")
+        for member in members:
+            data = self._http_get(f"/public/members/{member.id}")
+            member = json.load(data)
+            print(member)
+
+    def get_collection(self, id: str) -> Any:
+        return json.load(self._http_get(f"/public/collections/{id}"))
+
+    def get_collections(self) -> Iterable[Collection]:
+        data = self._http_get(f"/public/collections")
+        collections = json.load(data)
+
+        for collection in collections["data"]:
+            group_accesses: Optional[Tuple[GroupCollectionAccess, ...]] = None
+            member_accesses: Optional[Tuple[MemberCollectionAccess, ...]] = None
+            collection_data = self.get_collection(collection["id"])
+
+            group_accesses_data = tuple(sorted(self.get_collection_groups(collection_data["groups"])))
+
+            if group_accesses_data:
+                group_accesses = group_accesses_data
+                member_accesses_data = tuple(sorted(self.get_collection_members(collection_data["groups"])))
+                if member_accesses_data:
+                    member_accesses = member_accesses_data
+
+            yield Collection(
+                id=collection["id"],
+                external_id=collection["externalId"],
+                member_access=member_accesses,
+                group_access=group_accesses,
+            )
+
+    def get_group_members(self, id: str, name: str) -> Iterable[GroupMember]:
+        members = json.load(self._http_get(f"/public/groups/{id}/member-ids"))
+
+        for member in members:
+            member = json.load(self._http_get(f"/public/members/{member}"))
+            yield GroupMember(
+                member_id=member["id"],
+                member_name=member["name"],
+                group_name=name,
+            )
+
+    def set_member_type(self, type_id: int) -> str:
+        match type_id:
+            case 0:
+                type="owner"
+            case 1:
+                type="admin"
+            case 2:
+                type="user"
+            case 3:
+                type="manager"
+            case 4:
+                type="custom"
+        return type
```

Comment:
First we can define an enum like so:

```python
class MemberType(Enum):
    OWNER = 0
    ADMIN = 1
    USER = 2
    MANAGER = 3
    CUSTOM = 4
```

And then this function becomes trivial:

```suggestion
    def set_member_type(self, type_id: int) -> MemberType:
        return MemberType(type_id)
```

This method does not set the type though! `NamedTuple`s are immutable, you cannot change their fields. If you want to change a field, you can use [`_replace`](https://docs.python.org/3/library/collections.html#collections.somenamedtuple._replace). It does not mutate the thing on which you call it, it returns a _new_ value that is a copy of the original, but with some fields replaced.

---

If you don’t want this immutability, you can define a regular `class`, not inherit from `NamedTuple`, and then assign all fields in `__init__`, or more recently you can define a class and specify the fields like you do with `NamedTuple`, and then put [`@dataclass`](https://docs.python.org/3/library/dataclasses.html?highlight=dataclass#dataclasses.dataclass) on it. `@dataclass(frozen=True)` defines an immutable data class, which is pretty much the same as `NamedTuple` aside from some technicalities, but the default value for `frozen` is `False` and then it defines a class with mutable fields.

It’s a bit a matter of preference whether you make things more mutable (what you would do in a more imperative style) or immutable (what you do in a more functional style), personally I usually prefer to make things immutable.

-----

But anyway, this method is on `BitwardenClient` which does not have a `type` field anyway, I think the `set` is a misnomer and also this doesn’t need to be a method, it could just be a top-level function? (Or a `@staticmethod` on the class.) But if we go with the enum, then converting between `int` and `MemberType` becomes so easy that we can just delete this method entirely and write `MemberType(type_int)` in the place where we need to convert.

Reply:
Thanks for the explanation, this is gold for me !

I don't understand how this should work. In case of `get_members` the value return by the API is a `int` then we can return the correct permission for `MemberType` based on the `type_int` value.

In the context of `from_toml_dict` method for the `Member` class, we don't get the `type` from the configuration file as `int` but as a `string` to be user friendly with `owner|user...`. The values in `MemberType` are not matching, will not resolve the name and throw an error.

I set this helper method on `BitwardenClient` because it was only used to map the `int` type returned by the API with the `get_members` method to a `string` type that is returned by the configuration file.

Reply:
Aah yes I see, we want the enum values to be strings for the toml file, but integers for mapping the API responses. We need to pick one of the two for the enum values. Then what we can do for the other one is use a dict instead. For example:

```python
int_to_member_type: Dict[int, MemberType] = {
    0: MemberType.OWNER,
    1: MemberType.ADMIN,
    2: MemberType.USER,
    3: MemberType.MANAGER,
    4: MemberType.CUSTOM,
}
```

Now we can write `int_to_member_type[type_id]` to convert the integers that we get from the API.